src/query/service/src/interpreters/common/refresh_aggregating_index.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright 2021 Datafuse Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use common_base::runtime::GlobalIORuntime;
+use common_catalog::catalog::CatalogManager;
+use common_catalog::table_context::TableContext;
+use common_exception::Result;
+use common_meta_app::schema::IndexMeta;
+use common_meta_app::schema::ListIndexesByIdReq;
+use common_pipeline_core::Pipeline;
+use common_sql::plans::RefreshIndexPlan;
+use common_sql::BindContext;
+use common_sql::Binder;
+use common_sql::Metadata;
+use common_sql::NameResolutionContext;
+use log::info;
+use parking_lot::RwLock;
+use storages_common_table_meta::meta::Location;
+
+use crate::interpreters::Interpreter;
+use crate::interpreters::RefreshIndexInterpreter;
+use crate::pipelines::executor::ExecutorSettings;
+use crate::pipelines::executor::PipelineCompleteExecutor;
+use crate::sessions::QueryContext;
+
+pub struct RefreshAggIndexDesc {
+    pub catalog: String,
+    pub database: String,
+    pub table: String,
+    pub table_id: u64,
+}
+
+pub async fn hook_refresh_agg_index(
+    ctx: Arc<QueryContext>,
+    pipeline: &mut Pipeline,
+    desc: RefreshAggIndexDesc,
+) -> Result<()> {
+    if pipeline.is_empty() {
+        return Ok(());
+    }
+
+    if ctx
+        .get_settings()
+        .get_enable_sync_refresh_aggregating_index()?
+    {
+        pipeline.set_on_finished(move |err| {
```

Comment:
This maybe a issue if the pipeline already have `set_on_finished`.

Reply:
```
   pub fn set_on_finished<F: FnOnce(&Option<ErrorCode>) -> Result<()> + Send + Sync + 'static>(
        &mut self,
        f: F,
    ) {
        if let Some(on_finished) = self.on_finished.take() {
            self.on_finished = Some(Box::new(move |may_error| {
                on_finished(may_error)?;
                f(may_error)
            }));

            return;
        }

        self.on_finished = Some(Box::new(f));
    }
```

The `set_on_finished` may should return error if we have already `set_on_finished`, instead use the first set one.  cc @zhang2014

Reply:
Exactly this is a potential issue, the current pipeline implementation does not support multiple `set_on_finished` or something like `finish chain`.
Currently, we need manually check if the worked-on pipeline has a `set_on_finished`, if so, we need to modify the code in the existing `set_on_finished`.

Reply:
set_on_finished is chain.